### ISSUE TITLE
nodenext doesn't pick up on addition of `type` field to `package.json`

### ISSUE THREAD

**Issue by DanielRosenwasser** [microsoft/TypeScript#50086]
Here's a quick way you can clone this repro locally

```sh
https://github.com/DanielRosenwasser/daniel-ts-repros/
cd daniel-ts-repros
git checkout nodenext-package-watch-bug
```

______

```ts
// src/fileA.ts
import { foo } from "./fileB.mjs";

foo();
```

```ts
// src/fileB.mts
export function foo() {
}
```

```json5
// src/tsconfig.json
{
  "compilerOptions": {
    "target": "es2016",
    "module": "Node16",
    "outDir": "../out"
  }
}
```

```json5
// package.json

{
  "name": "nodenext-package-watch-bug",
  "version": "1.0.0",
  "description": "Repro",
  "author": "Daniel"
}
```

Currently this produces the following result in the editor or in `--watch` mode:

```
Module './fileB.mjs' cannot be imported using this construct. The specifier only resolves to an ES module, which cannot be imported synchronously. Use dynamic import instead.ts
```

If you add a `"type": "module"` to the `package.json`, then the error will not go away until you restart `tsc` or TS Server.

```diff
-  "author": "Daniel"
+  "author": "Daniel",
+  "type": "module"
```

**Comment by sheetalkamat** [comment]
Looks like package json is read for file format rather than for resolution .. which is why it’s not watched ..

**Comment by fatcerberus** [comment]
Possible root cause of #50058?

**Comment by sheetalkamat** [comment]
Working on fixing this, but its not going to be 2 line fix..

**Comment by cyberwombat** [comment]
Not sure if same issue but with `nodenext` my code seems to transpile but shows errors in editor. I made a super simple sandbox [here](https://codesandbox.io/p/github/cyberwombat/ts-esm-test/draft/kind-sea?file=%2Findex.ts&workspace=%257B%2522activeFileId%2522%253A%2522cl9bh5ug20002lrhb6woealfc%2522%252C%2522openFiles%2522%253A%255B%255D%252C%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522gitSidebarPanel%2522%253A%2522COMMIT%2522%252C%2522sidekickItems%2522%253A%255B%257B%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522cl9bh6bqv000blrhb74jv6h3z%2522%252C%2522key%2522%253A%2522cl9bh6b2o00523b6fhlt7a3mv%2522%252C%2522isMinimized%2522%253Afalse%257D%255D%257D).  

Basically this code won't work w `moduleResolution' set to `nodenext`.
```
import Ajv from "ajv"

export const getClient = (): Ajv => { 
    return  new Ajv() 
}
```

My current workaround is to add dozens of lines such as `declare namespace 'ajv'` in my code for about half of my modules - presumably cjs ones.  Calling  - as in my example:
```
import Ajv from "ajv"`

export const getClient = (): Ajv.default => { 
    return  new Ajv.default() 
}
````
Makes the editor code go away but now code doesn't transpile correctly (results in things like ajv['default']['default'])
